pkg/controller/baremetalhost/host_state_machine.go

```diff
@@ -134,21 +129,39 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
-func (hsm *hostStateMachine) shouldInitiateRegister(info *reconcileInfo) bool {
-	changeState := false
-	if hsm.Host.DeletionTimestamp.IsZero() {
-		switch hsm.NextState {
-		default:
-			changeState = !hsm.Host.Status.GoodCredentials.Match(*info.bmcCredsSecret)
-		case metal3v1alpha1.StateNone:
-		case metal3v1alpha1.StateRegistering, metal3v1alpha1.StateRegistrationError:
-		case metal3v1alpha1.StateDeleting:
-		}
+func (hsm *hostStateMachine) ensureRegistered(info *reconcileInfo) (result actionResult) {
+	if hsm.NextState == metal3v1alpha1.StateNone {
+		// We haven't yet reached the Registration state, so don't attempt
+		// to register the Host.
+		return
 	}
-	if changeState {
-		hsm.NextState = metal3v1alpha1.StateRegistering
+	if !hsm.Host.DeletionTimestamp.IsZero() {
+		// BUG(zaneb) We currently don't attempt to re-register the Host
+		// if we find it missing once a delete has been requested (in
+		// part because we are also willing to pass empty credentials
+		// after this point), but this means that we may not be able to
+		// deprovision a Host if the Ironic database pod is rescheduled
+		// during deprovisioning, or if the credentials need to be
+		// modified.
+		return
 	}
-	return changeState
+
+	if !hsm.Host.Status.GoodCredentials.Match(*info.bmcCredsSecret) {
```

Comment:
I think I would write this the opposite way, and use an early return here if the credentials do match. Then the bulk of the logic for the method is outside of any conditional block.

Reply:
You're right, that is definitely better. Fixed.